thrust/thrust/type_traits/logical_metafunctions.h

```diff
@@ -34,267 +26,84 @@
-{};
-
-template <typename T0, typename T1, typename T2, typename... TN>
-struct conjunction<T0, T1, T2, TN...> : std::conditional<T0::value, conjunction<T1, T2, TN...>, T0>::type
-{};
-
-/*! \endcond
- */
-#endif
-
-/*! \brief <tt>constexpr bool</tt> whose value is <tt>(... && Ts::value)</tt>.
- *
- *  \see conjunction
- *  \see conjunction_value
- *  \see <a href="https://en.cppreference.com/w/cpp/types/conjunction"><tt>std::conjunction</tt></a>
- */
-#if _CCCL_STD_VER >= 2014
-template <typename... Ts>
-constexpr bool conjunction_v = conjunction<Ts...>::value;
-#endif
-
-/*! \brief <a href="https://en.cppreference.com/w/cpp/types/integral_constant"><tt>std::integral_constant</tt></a>
- *  whose value is <tt>(... || Ts::value)</tt>.
- *
- *  \see disjunction_v
- *  \see disjunction_value
- *  \see <a href="https://en.cppreference.com/w/cpp/types/disjunction"><tt>std::disjunction</tt></a>
- */
-#if _CCCL_STD_VER >= 2017
-template <typename... Ts>
-using disjunction = std::disjunction<Ts...>;
-#else // Older than C++17.
-template <typename... Ts>
-struct disjunction;
-
-/*! \cond
- */
-
-template <>
-struct disjunction<> : std::false_type
-{};
-
-template <typename T>
-struct disjunction<T> : T
-{};
-
-template <typename T0, typename... TN>
-struct disjunction<T0, TN...> : std::conditional<T0::value != false, T0, disjunction<TN...>>::type
-{};
-
-/*! \endcond
- */
-#endif
-
-/*! \brief <tt>constexpr bool</tt> whose value is <tt>(... || Ts::value)</tt>.
- *
- *  \see disjunction
- *  \see disjunction_value
- *  \see <a href="https://en.cppreference.com/w/cpp/types/disjunction"><tt>std::disjunction</tt></a>
- */
-#if _CCCL_STD_VER >= 2014
-template <typename... Ts>
-constexpr bool disjunction_v = disjunction<Ts...>::value;
-#endif
-
-/*! \brief <a href="https://en.cppreference.com/w/cpp/types/integral_constant"><tt>std::integral_constant</tt></a>
- *  whose value is <tt>!Ts::value</tt>.
- *
- *  \see negation_v
- *  \see negation_value
- *  \see <a href="https://en.cppreference.com/w/cpp/types/negation"><tt>std::negation</tt></a>
- */
-#if _CCCL_STD_VER >= 2017
-template <typename T>
-using negation = std::negation<T>;
-#else // Older than C++17.
-template <typename T>
-struct negation;
-
-/*! \cond
- */
-
-template <typename T>
-struct negation : std::integral_constant<bool, !T::value>
-{};
-
-/*! \endcond
- */
-#endif
-
-/*! \brief <tt>constexpr bool</tt> whose value is <tt>!Ts::value</tt>.
- *
- *  \see negation
- *  \see negation_value
- *  \see <a href="https://en.cppreference.com/w/cpp/types/negation"><tt>std::negation</tt></a>
- */
-#if _CCCL_STD_VER >= 2014
-template <typename T>
-constexpr bool negation_v = negation<T>::value;
+//! \addtogroup utility
+//! \{
+//! \addtogroup type_traits Type Traits
+//! \{
+
+using ::cuda::std::conjunction;
+using ::cuda::std::disjunction;
+using ::cuda::std::negation;
+#if _CCCL_STD_VER >= 2014 && !defined(_LIBCUDACXX_HAS_NO_VARIABLE_TEMPLATES)
```

Comment:
Can it happen that we have C++14 but no variable templates?

Reply:
Yes, its mostly older gcc

Reply:
Discussed offline: Thrust previously offered the variable templates when just `_CCCL_STD_VER >= 2014` was true. However, some GCC versions claim C++14 support but do not properly support variable templates yet. This would have resulted in a compile error in the past.

Now, we just don't provide the variable templates, also resulting in a compile error if the compiler does not support that feature.